vpr/src/analytical_place/full_legalizer.cpp

```diff
@@ -281,6 +291,710 @@ static LegalizationClusterId create_new_cluster(PackMoleculeId seed_molecule_id,
+                    if (pack_status == e_block_pack_status::BLK_PASSED)
+                        fit_in_a_neighbor = true;
+                }
+
+                // Clean the new cluster to avoid increasing memory footprint.
+                cluster_legalizer.clean_cluster(new_cluster_id);
+
+                // Erase old cluster id and add new one.
+                it = clusters.erase(it);
+                clusters.insert(new_cluster_id);
+            }
+            // Stop iterating neighbor tiles if current molecule already fit in a neighbor cluster.
+            if (fit_in_a_neighbor) {
+                mols_clustered.insert(molecule_id);
+                break;
+            }
+        }   
+    }
+}
+
+void FlatRecon::orphan_window_clustering(ClusterLegalizer& cluster_legalizer,
+                                         const vtr::vector<LogicalModelId, std::vector<t_logical_block_type_ptr>>& primitive_candidate_block_types,
+                                         std::unordered_set<LegalizationClusterId>& created_clusters,
+                                         int search_radius) {
+    std::string timer_label = "Orphan Window Clustering with search radius " + std::to_string(search_radius);
+    vtr::ScopedStartFinishTimer orphan_window_clustering(timer_label);
+
+    // Create unclustered blocks spatial data. It stores vector of molecules ids
+    // for each tile location of [layer][x][y].
+    auto [layer_num, width, height] = device_grid_.dim_sizes();
+    vtr::NdMatrix<std::unordered_set<PackMoleculeId>, 3> unclustered_tile_molecules({layer_num, width, height});
+    std::vector<PackMoleculeId> unclustered_blocks;
+    for (APBlockId blk_id: ap_netlist_.blocks()) {
+        PackMoleculeId mol_id = ap_netlist_.block_molecule(blk_id);
+        if (cluster_legalizer.is_mol_clustered(mol_id))
+            continue;
+        t_physical_tile_loc tile_loc = mol_desired_physical_tile_loc[mol_id];
+        unclustered_tile_molecules[tile_loc.layer_num][tile_loc.x][tile_loc.y].insert(mol_id);
+        unclustered_blocks.push_back(mol_id);
+    }
+
+    // Sort unclustered blocks by highest external input pins.
+    std::sort(unclustered_blocks.begin(), unclustered_blocks.end(),
+            [this](const PackMoleculeId& a, const PackMoleculeId& b) {
+                int ext_pins_a = prepacker_.calc_molecule_stats(a, atom_netlist_, arch_.models).num_used_ext_inputs;
+                int ext_pins_b = prepacker_.calc_molecule_stats(b, atom_netlist_, arch_.models).num_used_ext_inputs;
+                return ext_pins_a > ext_pins_b;
+            });
+
+    for (PackMoleculeId seed_mol_id: unclustered_blocks) {
+        if (cluster_legalizer.is_mol_clustered(seed_mol_id))
+            continue;
+
+        // Start the new cluster with seed molecule using full strategy.
+        cluster_legalizer.set_legalization_strategy(ClusterLegalizationStrategy::FULL);
```

Comment:
Could comment that this could waste time vs. using the fast strategy first and falling back to full, but currently orphan clustering doesn't take that long as few molecules are clustered.